cores/esp32/esp32-hal-uart.c

```diff
@@ -45,73 +45,126 @@ struct uart_struct_t {
-    {(volatile uart_dev_t *)(DR_REG_UART1_BASE), 1, NULL, NULL},
-    {(volatile uart_dev_t *)(DR_REG_UART2_BASE), 2, NULL, NULL}
+    {(volatile uart_dev_t *)(DR_REG_UART_BASE), 0, NULL, NULL,-1,false},
+    {(volatile uart_dev_t *)(DR_REG_UART1_BASE), 1, NULL, NULL,-1,false},
+    {(volatile uart_dev_t *)(DR_REG_UART2_BASE), 2, NULL, NULL,-1,false}
 };
 #else
-#define UART_MUTEX_LOCK()    do {} while (xSemaphoreTake(uart->lock, portMAX_DELAY) != pdPASS)
-#define UART_MUTEX_UNLOCK()  xSemaphoreGive(uart->lock)
+#define UART_MUTEX_LOCK()    do {} while (xSemaphoreTakeRecursive(uart->lock, portMAX_DELAY) != pdPASS)
+#define UART_MUTEX_UNLOCK()  xSemaphoreGiveRecursive(uart->lock)
 
 static uart_t _uart_bus_array[3] = {
-    {(volatile uart_dev_t *)(DR_REG_UART_BASE), NULL, 0, NULL, NULL},
-    {(volatile uart_dev_t *)(DR_REG_UART1_BASE), NULL, 1, NULL, NULL},
-    {(volatile uart_dev_t *)(DR_REG_UART2_BASE), NULL, 2, NULL, NULL}
+    {(volatile uart_dev_t *)(DR_REG_UART_BASE), NULL, 0, NULL, NULL,-1,false},
+    {(volatile uart_dev_t *)(DR_REG_UART1_BASE), NULL, 1, NULL, NULL,-1,false},
+    {(volatile uart_dev_t *)(DR_REG_UART2_BASE), NULL, 2, NULL, NULL,-1,false}
 };
 #endif
 
-static void uart_on_apb_change(void * arg, apb_change_ev_t ev_type, uint32_t old_apb, uint32_t new_apb);
+//#define DEBUG_RX
+
+static bool uart_on_apb_change(void * arg, apb_change_ev_t ev_type, uint32_t old_apb, uint32_t new_apb);
+
+static void _uart_rx_read_fifo( uart_t * uart){
+    uint8_t c;
+    UART_MUTEX_LOCK();
+    UBaseType_t count=0, spaces=0;
+    // wr_addr, rd_addr are circular buffer pointers, wrapping is controlled by UART config,
+    // standard fifo is 128 byte, but address is 11bits (1024byte buffer, allocated into 6 128byte buffers rx/tx)
+    
+    count = (uart->dev->mem_rx_status.wr_addr - uart->dev->mem_rx_status.rd_addr)% 0x80;
+    if(uart->queue != NULL) {
+        spaces=uxQueueSpacesAvailable( uart->queue);
+    
+        while((spaces >0)&&(count-- > 0)){
+        //(uart->dev->status.rxfifo_cnt || (uart->dev->mem_rx_status.wr_addr != uart->dev->mem_rx_status.rd_addr))) {
+            c = uart->dev->fifo.rw_byte;
+            xQueueSend(uart->queue, &c,0);
+            spaces--;
+        }
+    }  
+    if(spaces == 0){ // disable rxFifo_full interrupt, no room in Queue or queue does not exit
+        uart->dev->int_ena.rxfifo_full = 0;
+        uart->dev->int_ena.rxfifo_tout = 0;
+#ifdef DEBUG_RX
+        digitalWrite(25,HIGH);
+#endif
+    }
+    UART_MUTEX_UNLOCK();
+}
 
 static void IRAM_ATTR _uart_isr(void *arg)
 {
     uint8_t i, c;
-    BaseType_t xHigherPriorityTaskWoken;
-    uart_t* uart;
+    BaseType_t xHigherPriorityTaskWoken=pdFALSE;
+    uart_t* uart =(uart_t *)arg;
+    if( uart != NULL){
+        UBaseType_t spaces=0,count=0;
+        count = (uart->dev->mem_rx_status.wr_addr - uart->dev->mem_rx_status.rd_addr)%0x80;
+        if(uart->queue != NULL) {
+            spaces=uxQueueSpacesAvailable( uart->queue);
```

Comment:
`uxQueueSpacesAvailable` ok to be called from isr?

Reply:
I can't find any documentation either way.